### PR TITLE
[assoc_scan] Additional_inputs gradients

### PR DESCRIPTION
**Pull request by bohnstingl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #192971
* #192654
* #192389

This PR introduces support for additional inputs of associative_scan.
Such inputs may arise from user-defined function, as well as from
lifted parameters, such as ints or SymInts, during compilation.

cc @ydwu4

### CODE REVIEW

File: torch/_higher_order_ops/associative_scan.py [hunk 1]
```diff
@@ -699,7 +711,10 @@ def backward(ctx, *gl_ys):
             return (
                 *[None] * 3,
                 *[torch.zeros_like(x) for x in xs],
-                *[None] * num_additional_inputs,
+                *[
+                    torch.zeros_like(a) if m else None
+                    for m, a in zip(additional_inputs_tensor_mask, additional_inputs)
```
Comment (Skylion007) [inline torch/_higher_order_ops/associative_scan.py]: Add strict=True?

**Comment by bohnstingl** [comment]
@claude review these changes

**Comment by bohnstingl** [comment]
@claude Please check whether the comments from https://github.com/pytorch/pytorch/pull/192971#discussion_r3759475437 and from https://github.com/pytorch/pytorch/pull/192971#issuecomment-5297917021 have been resolved with the latest commits

**Comment by bohnstingl** [comment]
@ydwu4 This PR is ready for review.
